llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp

```diff
@@ -6872,35 +6883,6 @@ void OpenMPIRBuilder::loadOffloadInfoMetadata(StringRef HostFilePath) {
   loadOffloadInfoMetadata(*M.get());
 }
 
-Function *OpenMPIRBuilder::createRegisterRequires(StringRef Name) {
```

Comment:
Removing this function breaks the compilation of the OpenMP MLIR dialect, as it's used there (mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp: `convertRequiresAttr()`).

My understanding is that creating this function would no longer be necessary, so the solution to that should be to remove `convertRequiresAttr()` and replace the call to it in `OpenMPDialectLLVMIRTranslationInterface::amendOperation()` to `return success()` in the same file. Flang should already pick up your other changes, so REQUIRES information should still work in Fortran after this.

Reply:
Thanks for the heads up. Do you know if there will be any other changes required to make the requires information use the "new" format?

Reply:
Flang already picks up your changes to `createOffloadEntriesAndInfoMetadata()` and it appears to be putting the corresponding flags into a structure, though I'm not sure it's the right one.

I ran `flang-new -fc1 -emit-llvm -fopenmp flang/test/Lower/OpenMP/requires.f90 -o -` for a simple test with a single target region and I'm seeing two offloading entries being produced, which I'm not sure whether it's expected behavior. It looks like only one of them has the requires flags and the other one is the one that's properly linked with the kernel.

```llvmir
@.omp_offloading.entry_name = internal unnamed_addr constant [43 x i8] c"__omp_offloading_10307_d2a215c__QQmain_l12\00"
@.omp_offloading.entry.__omp_offloading_10307_d2a215c__QQmain_l12 = weak constant %struct.__tgt_offload_entry { ptr @.__omp_offloading_10307_d2a215c__QQmain_l12.region_id, ptr @.omp_offloading.entry_name, i64 0, i32 0, i32 0 }, section "omp_offloading_entries", align 1

@.omp_offloading.entry_name.1 = internal unnamed_addr constant [1 x i8] zeroinitializer
@.omp_offloading.entry. = weak constant %struct.__tgt_offload_entry { ptr null, ptr @.omp_offloading.entry_name.1, i64 0, i32 22, i32 10 }, section "omp_offloading_entries", align 1

...
%14 = call i32 @__tgt_target_kernel(ptr @1, i64 -1, i32 -1, i32 0, ptr @.__omp_offloading_10307_d2a215c__QQmain_l12.region_id, ptr %kernel_args)
```

Reply:
The test itself has "requires unified_shared_memory reverse_offload" (it originally just checks that the flags are passed to MLIR), so that's where the "10" is coming from. The "22" I don't know what it's supposed represent.

Reply:
I encoded the fact that this is a "requires" entry into the global with `16`. Realistically it shouldn't make a difference in the runtime since as long as `16` is set and the address is null it will get counted as a requires entry. It's just a little weird that it wouldn't *just* be 16. I'll double check. Changing these entries is also on the list.

Reply:
It was a very obvious problem. I mixed up hexadecimal and did `0x16` when I meant either `0x10` or `16`. Fixed now.

Reply:
> It was a very obvious problem. I mixed up hexadecimal and did `0x16` when I meant either `0x10` or `16`. Fixed now.

Oh, then the "16" bit just happened to be part of the number that was being set. Glad to hear this wasn't some unexpected interaction specific to Flang.